### PR TITLE
Don't block FRP network if writing to stdin blocks

### PR DESCRIPTION
**Pull request by meditans**
This PR solves https://github.com/reflex-frp/reflex-process/issues/9

The way this was done is adding a `TChan` to act as an unbounded buffer between the FRP network and the `stdin` handle. There's also a test suite that checks for this property. I suggest running the test suite with:
```
nix-shell -E '((import ./reflex-platform {}).ghc.callCabal2nix "reflex-process" ./. {}).env' --run 'cabal test --test-show-details=streaming'
```
so that one can check both the output of the `cat` process, and the eventual messages that arise from a blocking process (for now, I have a broken pipe that arises from the way I'm handling the finally clause in the test suite).

### CODE REVIEW

File: src/Reflex/Process.hs
Comment (3noch) [inline src/Reflex/Process.hs]: You can just call this `_processConfig_createProcess`

Comment (meditans) [reply src/Reflex/Process.hs]: Done, thanks!

File: src/Reflex/Process.hs
Comment (tomsmalley) [inline src/Reflex/Process.hs]: Leftover

**Review by tomsmalley — DISMISSED** [review]
Seems reasonable, but we should move the headless host stuff elsewhere.

File: src/Reflex/Process.hs
Comment (tomsmalley) [reply src/Reflex/Process.hs]: Oh, I missed the later commit. Nevermind

**Comment by lspitzner** [comment]
I am mildly certain that `TChan` is overkill. `Control.Concurrent.Chan` would do, ~~and even `Data.IORef.IORef` using `atomicModifyIORef` would work.~~ Afaik, `IORef [a]` serves fine as a channel as long as you have a single consumer that can always drain all elements (EDIT: _and_ the consumer must not need to block, which this code _does_ need.) ~~Which you could do here, even though it adds a bit of boilerplate.~~

~~I'll let the maintainers of this decide whether they want the IORef performance or the `Chan` maintainability.~~

**Comment by lspitzner** [comment]
Oh wait, IORef cannot block. Whoops. Needs `Chan`.

**Comment by 3noch** [comment]
@lspitzner The only reason we went with `TChan` is it affords a slightly nicer API if you need it, but seeing that we're not exposing the channel itself it's probably moot.

**Comment by 3noch** [comment]
@lspitzner I think we should actually expose the `TChan` which makes me prefer `TChan` over `Chan` because `TChan` has a much richer API, particularly if you hope to build a pressure relief valve since it is unbounded.

**Comment by lspitzner** [comment]
true, I approve.

File: src/Reflex/Process.hs
Comment (3noch) [inline src/Reflex/Process.hs:42]: Let's move the `error` call into this default so that this type signature is more precise.
```suggestion
    :: P.CreateProcess -> IO (Handle, Handle, Handle, ProcessHandle)
```

File: src/Reflex/Process.hs
Comment (3noch) [inline src/Reflex/Process.hs]: Let's document as an `NB:` that this uses an *unbounded* channel to prevent the networking from blocking.

File: src/Reflex/Process.hs
Comment (3noch) [inline src/Reflex/Process.hs]: I don't think we want to use `Char8` here even though I'm not sure there's actually any error. But we definitely don't want to be shedding any doubt about whether this is suitable for arbitrary bytes.

**Comment by 3noch** [comment]
Then again, just abstracting this a bit more makes the choice of channel a trivial configuration. Will go that route instead.

File: src/Reflex/Process.hs
Comment (3noch) [reply src/Reflex/Process.hs]: Oh my I just realized this is using `hPutStrLn` which is a bug anyway. That's going to insert newlines in random places. We should leave the line-breaking to the input stream itself.

File: src/Reflex/Process.hs
Comment (3noch) [inline src/Reflex/Process.hs]: heh whoops.

File: src/Reflex/Process.hs
Comment (3noch) [reply src/Reflex/Process.hs:42]: Actually I'm having second thoughts about this because of how we use `createProcess` below.

**Review by 3noch — APPROVED** [review]

**Pull request merged**
